teps/0011-redirecting-step-inputs-and-outputs.md

```diff
@@ -0,0 +1,143 @@
+---
+title: redirecting-step-inputs-and-outputs
+authors:
+  - "@chhsia0"
+creation-date: 2020-08-17
+last-updated: 2020-08-17
+status: proposed
+---
+
+# TEP-0011: Redirecting Step Inputs and Outputs
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Requirements](#requirements)
+- [Proposal](#proposal)
+    - [Example Usage](#example-usage)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+Consuming outputs of a step in another step is a common pattern in writing Tasks. However, this is currently tedious to do. Task authors have to overwrite the image entrypoint with either `sh -c` or the `script` field to wrap up the command to run with an explicit I/O redirection. This is not even possible if the image does not come with a shell, e.g., images built by [`ko`](https://github.com/google/ko).
+
+To achieve the functionality of I/O redirection between steps and even Tasks, Tekton will add new fields to steps for Task authors to specify paths to store stdout/stderr to and read stdin from.
+
+## Motivation
+
+Allow Task authors to support the following use cases:
+
+* Run image `gcr.io/k8s-staging-boskos/boskosctl` with args in a step and process its output through another `jq` image in a subsequent step to acquire the project name without overwriting the image entrypoint, building custom image with both utilities, or looking into container logs ([pipeline#2925](https://github.com/tektoncd/pipeline/issues/2925)).
+
+* Run image `alpine/git` with args `["describe", "--tags"]` in a step and extract the stdout as a Task result without overwriting the image entrypoint to be used by subsequent tasks.
+
+* In general, enable Task authors to access stdout/stderr of a step easily in another step in the same or different Task, without requiring authors to: 1) overwrite image entrypoints; 2) write wrapping scripts solely to do I/O redirection; and 3) build wrapper images to include all needed utilities.
```

Comment:
This would be achieveable by having Tekton always write previous steps' stdout/stderrr to some well-known documented path, e.g., `/tekton/steps/2/stdout` or something similar (possibly also symlinked at `/tekton/steps/[step-name]/stdout`)

Reply:
I'm a bit hesitate to always copy stdout/stderr, because sometimes it could be large can fill the disk. In the current design we require users to "hint" Tekton to capture stdout/stderr for other commands to consume, which tends to be small, or at least users are aware that they're capturing something large.

Reply:
Yeah, I would definitely want to avoid *always* copy stdout/stderr, it's gonna add a lot of I/O for nothing. Enabling the copy explicitly seems way better.

Reply:
Also the stream duplication itself might fail (if an error happens in the underlying OS/kernel system call), so avoiding such additional syscall if not necessary sounds a bit better.

Reply:
Makes sense. wdyt about a single boolean flag to signal that the user wants stdout/stderr written to the conventional path?

Reply:
Adding this as an option in the proposal.